cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,440 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include <linalg/binary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <memory>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <common/cumlHandle.hpp>
+#include <cuML.hpp>
+
+namespace MLCommon {
+namespace Matrix {
+
+__global__ void identity_matrix_kernel(double** I, int r);
+
+template <typename T>
+void cudaFreeT(T* ptr) {
+  CUDA_CHECK(cudaFree(ptr));
+}
+
+// https://stackoverflow.com/questions/32685540/why-cant-i-compile-an-unordered-map-with-a-pair-as-key
+struct pair_hash {
+  template <class T1, class T2>
+  std::size_t operator()(const std::pair<T1, T2>& p) const {
+    auto h1 = std::hash<T1>{}(p.first);
+    auto h2 = std::hash<T2>{}(p.second);
+
+    // Not the best pair_hash, but it works for now...
+    return 100 * h1 + h2;
+  }
+};
+
+//! An allocation function for `BatchedMatrixMemory`. Written as a free function because I had trouble
+//! getting the __device__ lambda to compile as a member function of the `BatchedMatrixMemory`
+//! struct.
+std::pair<double*, double**> BMM_Allocate(
+  std::pair<int, int> shape, int num_batches, bool setZero,
+  std::shared_ptr<ML::deviceAllocator> allocator) {
+  int m = shape.first;
+  int n = shape.second;
+
+  // Allocate dense batched matrix and possibly set to zero
+  double* A_dense =
+    (double*)allocator->allocate(sizeof(double) * m * n * num_batches, 0);
+  if (setZero)
+    CUDA_CHECK(cudaMemset(A_dense, 0, sizeof(double) * m * n * num_batches));
+
+  // allocate and fill array of pointers to each batch matrix
+  double** A_array =
+    (double**)allocator->allocate(sizeof(double*) * num_batches, 0);
+  // fill array of pointers to each batch matrix.
+  auto counting = thrust::make_counting_iterator(0);
+  thrust::for_each(counting, counting + num_batches, [=] __device__(int bid) {
+    A_array[bid] = &(A_dense[bid * m * n]);
+  });
+  return std::make_pair(A_dense, A_array);
+}
+
+//! The BatchedMatrix class provides storage and a number of linear operations on collections (batches) of matrices of identical shape.
+class BatchedMatrix {
+ public:
+  //! Constructor that allocates memory using the memory pool.
+  //! @param m Number of rows
+  //! @param n Number of columns
+  //! @param num_batches Number of matrices in the batch
+  //! @param pool The memory pool
+  //! @param setZero Should matrix be zeroed on allocation?
+  BatchedMatrix(int m, int n, int num_batches, cublasHandle_t cublasHandle,
+                std::shared_ptr<ML::deviceAllocator> allocator,
+                cudaStream_t stream = 0, bool setZero = true)
```

Comment:
please do not have a default value for cuda streams inside ml-prims. Callers need to explicitly pass this argument.